### PR TITLE
Added support for Singularity

### PR DESCRIPTION
**Pull request by sandepat**
1. Updated the wrapper script to link singularity to CVMFS apptainer image (to execute singularity commands)
2. CVMF apptainer image - /cvmfs/dirac.egi.eu/container/apptainer/bin/apptainer

### CODE REVIEW

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 1]
```diff
@@ -19,6 +19,23 @@ function error {
...
+function function_singularity {
+    cat > singularity << 'EOF'
+          #!/bin/bash
+          /cvmfs/dirac.egi.eu/container/apptainer/bin/apptainer "$@"
+EOF
+}
+
+# Call the function to create the singularity script and make it executable
+function_singularity
+chmod +x singularity
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:31]: as for udocker, better do a `chmod a+x`.
also, move this line inside the `function_singularity` method to change the rights just after you create it

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 2]
```diff
@@ -19,6 +19,23 @@ function error {
   echo [ ERROR - $D ] $* >&2
 }
 
+function function_singularity {
+    cat > singularity << 'EOF'
+          #!/bin/bash
+          /cvmfs/dirac.egi.eu/container/apptainer/bin/apptainer "$@"
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:25]: Not a good idea to have an hardcoded path like that. We will use a environment variable for the apptainer location, like `APPTAINER_LOCATION`.

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 3]
```diff
@@ -19,6 +19,23 @@ function error {
...
+EOF
+}
+
+# Call the function to create the singularity script and make it executable
+function_singularity
+chmod +x singularity
+export APPTAINER_PULLFOLDER="/cvmfs/dirac.egi.eu/container/apptainer/"
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:32]: Is it the env variable that apptainer uses to find images ?

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 4]
```diff
@@ -19,6 +19,23 @@ function error {
...
+
+# Define a function to run singularity commands
+singularity() {
+    ./singularity "$@"
+}
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:37]: It is better to add `PWD` to the `$PATH` like it's done for udocker

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 5]
```diff
@@ -19,6 +19,23 @@ function error {
...
 
+function function_singularity {
+    cat > singularity << 'EOF'
+          #!/bin/bash
+          /cvmfs/dirac.egi.eu/container/apptainer/bin/apptainer "$@"
+EOF
+}
+
+# Call the function to create the singularity script and make it executable
+function_singularity
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:30]: We have the choice to always add singularity/apptainer (like you did), or to do it only when the boutiques app uses singularity. (like it's done for udocker).
@camarasu what do you prefer ?

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 1 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm:31]: Updated to chmod a+x singularity

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 3 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm:32]: yes. Should this also be added to the configuration file?

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 2 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm:25]: Done. The path for the $APPTAINER_LOCATION should be declared in gaswConfiguration?

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 4 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm:37]: done.

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 5 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm]: done

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java [hunk 6]
```diff
@@ -104,6 +111,15 @@ public BoutiquesApplication parseApplication(String descriptor) throws InvalidBo
                         exception);
             }
         }
+        // Custom property
+        JSONObject customObject = getObjectValue(parsedDescriptor, "custom", true);
+        if (customObject != null) {
+            try {
+                application.setVipContainer(getStringValue(customObject, "vip:imagepath", true));
+            } catch (InvalidBoutiquesDescriptorException exception) {
```
Comment (axlbonnet) [inline vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java:119]: i think we can avoid this exception catching, as the key is optional and the new exception does not add useful information to the original one.

File: vip-portal/src/main/resources/vm/wrapper.vm [hunk 7]
```diff
@@ -187,7 +203,12 @@ TMP_FOLDER=/tmp/$TMP_FOLDER
 TMP_FOLDER=../$TMP_FOLDER
 #end
 mkdir -p $TMP_FOLDER
+
+#if($imagepath == "")
```
Comment (axlbonnet) [inline vip-portal/src/main/resources/vm/wrapper.vm:207]: does it work if `imagepath` is null ?

**Review by axlbonnet — COMMENTED** [review]
almost ready !
Just a minor improvement and a interogation.

File: vip-portal/src/main/resources/vm/wrapper.vm (hunk 7 above)
Comment (sandepat) [reply vip-portal/src/main/resources/vm/wrapper.vm:207]: If $imagepath is null or undefined in, the #if condition will handle it as if it were an empty string (""). 
In VTL, a null value and an empty string are treated similarly in conditions. So I think the statement handles both conditions appropriately.

File: vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java (hunk 6 above)
Comment (sandepat) [reply vip-application/src/main/java/fr/insalyon/creatis/vip/application/client/view/boutiquesParsing/BoutiquesParser.java]: Done

**Comment by camarasu** [comment]
Well done, thank you!

**Pull request merged**